bgpd/bgp_bmp.c

```diff
@@ -265,34 +468,61 @@ static inline int bmp_get_peer_type_vrf(vrf_id_t vrf_id)
 }
 
 /* determine the peer type for per-peer headers from a struct peer
- * provide a bgp->peer_self for loc-rib
+ * bgp->peer_self will NOT give loc-rib peer type
  */
 static inline int bmp_get_peer_type(struct peer *peer)
 {
-	if (peer->bgp->peer_self == peer)
-		return BMP_PEER_TYPE_LOC_RIB_INSTANCE;
-
 	return bmp_get_peer_type_vrf(peer->bgp->vrf_id);
 }
 
-static inline int bmp_get_peer_distinguisher(struct bmp *bmp, afi_t afi,
-					     uint8_t peer_type,
+/* compute peer distinguisher from bmp session, afi and peer_type
+ * store it in result_ref
+ *
+ * providing AFI_UNSPEC as afi value will use any RD configured in the VRF
+ *
+ * returns	1 on error that needs message discarding
+ *		0 if successful
+ */
+static inline int bmp_get_peer_distinguisher(struct bgp *bgp, afi_t afi,
 					     uint64_t *result_ref)
 {
-
-	/* remove this check when the other peer types get correct peer dist.
-	 *(RFC7854) impl.
-	 * for now, always return no error and 0 peer distinguisher as before
-	 */
-	if (peer_type != BMP_PEER_TYPE_LOC_RIB_INSTANCE)
-		return (*result_ref = 0);
-
-	/* sending vrf_id or rd could be turned into an option at some point */
-	struct bgp *bgp = bmp->targets->bgp;
-
 	/* vrf default => ok, distinguisher 0 */
 	if (bgp->inst_type == VRF_DEFAULT)
-		return (*result_ref = 0);
+		return (int)(*result_ref = 0);
+
+	/* if requested afi has no rd configured find any other */
+	if (afi >= AFI_MAX || !CHECK_FLAG(bgp->vpn_policy[afi].flags,
+			BGP_VPN_POLICY_TOVPN_RD_SET))
+		afi = AFI_UNSPEC;
+
+	/* afi not known, use any afi configured in this vrf */
+	if (afi == AFI_UNSPEC) {
+		{ /* scope lock iter variables */
+			afi_t afi_rd_lookup;
+			safi_t _;
+
+			FOREACH_AFI_SAFI (afi_rd_lookup, _) {
+
+				if (CHECK_FLAG(bgp->vpn_policy[afi_rd_lookup]
+						       .flags,
+					       BGP_VPN_POLICY_TOVPN_RD_SET)) {
+					afi = afi_rd_lookup;
+					/* stop FOREACH_AFI_SAFI macro */
+					afi_rd_lookup = AFI_MAX;
+				}
+
+				/* break safi sub-loop to go over next afi */
+				break;
+			}
+		}
+
+		/* no RD found for any AFI => error => skip message */
+		if (afi == AFI_UNSPEC) {
+			zlog_warn(
+				"skipping bmp message for reason: can't get peer distinguisher (no RD configured)");
+			return 1;
```

Comment:
please no log message when returning 1. it will be called too many times if the vrf did not yet go  up.

Reply:
okay but how should we communicate this issue then? it will be hard to troubleshoot without logs

Reply:
1) about bmp messages, I suggest you use a debug condition before displaying it. (debug bgp bmp ?)
2) about the error case, actually, not having RD may not be an issue.
this is what happens when you configure BMP on a VRF instance.
I hesitate between:
- maintain the peer type to RD instance peer, and fallback to RD value = 0:<vrf_id>.
- move the peer type to Local Instance Peer, and use RD value = 0:<vrf_id>.

the only case where the message will not be sent, will be when VRF_ID is unknown.

Reply:
i would prefer the 2nd approach since it would avoid clashes for the RD instance peer where you have a vrf that fallbacks to 0:<vrf_id> and one that is actually configured with say RD == 0::2
do you want to include that in this PR as well ?